### PR TITLE
feat: Huggingface BaseTransformerTrial implementation

### PR DESCRIPTION
**Pull request by liamcli**
## Description
Current commit contains an implementation of a BaseTransformTrial class along with helper functions to enable users to easily customize a trial using huggingface transformers for a specific task/config.  There is one such example provided for named entity recognition under examples/huggingface/token-classification.

## RFC
* Please provide feedback on the API and helper functions.  

## Test Plan
- [x] Already tested with named entity recognition example as well as a few others not yet in a PR (e.g. language modeling, multiple-choice, text-classification).
- [x] Add tests for utility functions.

## Checklist
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.

### CODE REVIEW

File: examples/nlp/hf_trial/base/base_trial.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
+from typing import Dict, Sequence, Union
```
Comment (rb-determined-ai) [inline examples/nlp/hf_trial/base/base_trial.py:1]: Question: why not put this file in the main determined package?

File: examples/nlp/hf_trial/base/base_trial.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+                self.hparams["model_name"],
+                config=self.config,
+            )
+        self.model = self.context.wrap_model(model)
+
+        # Dataset needs to be build before we initialize optimizers because
+        # a pickle is performed but we can't serialize distributed optimizer.
+        # This also has to happen after the model is built because we need to
+        # know which fields the model needs in the forward step to remove
+        # unncessary fields from the data dict.
```
Comment (rb-determined-ai) [inline examples/nlp/hf_trial/base/base_trial.py:88]: question: why does building the dataset have anything to do with pickling the optimizer?

**Review by rb-determined-ai — COMMENTED** [review]
Ok, so I know that this resembles what I originally suggested in a lot of ways, but I the more I look at it, the more I think the way I originally suggested isn't really that great.

From a high-level perspective, we have a plugin system, and our users are used to libraries.  At the most basic level, the core difference between a plugin system and a library from the user's perspective is the direction of calls.   With a library, the users and control the entire time, and they decide when to call our code and how to call it.  With a plugin system, we maintain control the entire time, and we decide when and how to call the user code.  I think by and large, users prefer libraries to plugins.  Part of our challenge is to design a plugin system that lets the user feel like they are in control anyway.

From in API design perspective, I am concerned about the number of callbacks in this design.  This API looks to me like it is going to have many of the shortcomings that our PyTorchTrial had before the PyTorch flexible primitives project. That is, the more you want to customize the more callbacks you have to implement.  We had a lot of positive feedback on the PyTorchTrial after we reduced the number of callbacks, and I think we should start from there as we implement this.

Another concern that I have is the amount of behavior that is encoded in the base class. A very long time ago, we actually implemented all the trial controllers as base classes (like this [old PyTorchTrial](https://github.com/determined-ai/pedl/blob/v0.9.0/pedl/frameworks/pytorch/pytorch_trial.py). This strategy is a common pattern for user code, because it is easy to share code between parent and subclasses.  However that works well because users know their own code very intimately. Users are not intimately familiar with our code (the base class).  It is generally better practice to have a well defined interface and strong separation between what is ours and what belongs to the user.  This releases the user from having to study the actual implementation of the base class in order to use our system.

Is there a way that we can package up some or all of the `__init__` functionality and the `build_optimizer` functionality into either a few functions, or perhaps into a subclass of the PyTorchContext?  That way users can implement fewer callbacks.

File: examples/nlp/hf_trial/base/base_trial.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
+from typing import Dict, Sequence, Union
+
+import torch
+
+import numpy as np
+
+from transformers import (
```
[the remark this answers is not in the document]
Comment (liamcli) [reply examples/nlp/hf_trial/base/base_trial.py]: thanks for the explanation, that makes a lot of sense.

File: model_hub/examples/huggingface/token-classification/ner_trial.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+from seqeval import metrics
+import transformers
+import datasets
+
+from determined.pytorch import DataLoader, PyTorchTrialContext, LRScheduler
+from model_hub import huggingface as hf
+from model_hub import utils
+
+
+class NERTrial(hf.BaseTransformerTrial):
```
Comment (naren-determined) [inline model_hub/examples/huggingface/token-classification/ner_trial.py:18]: Could we separate model code from Determined code similar to the refactor PR?

File: model_hub/model_hub/huggingface/_arg_parser.py [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+@dataclass
+class ConfigKwargs:
+    pretrained_model_name_or_path: str = field(default=None)
+    cache_dir: Optional[str] = field(
+        default=None,
+        metadata={
+            "help": "Where do you want to store the pretrained models downloaded from huggingface.co"
+        },
+    )
+    revision: Optional[str] = field(
```
Comment (naren-determined) [inline model_hub/model_hub/huggingface/_arg_parser.py:52]: Can we title revision something else -- maybe `model_version`? This name is not apparent to me as to what it means.

File: model_hub/model_hub/huggingface/_arg_parser.py [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+            "help": "Where do you want to store the pretrained models downloaded from huggingface.co"
+        },
+    )
+    use_fast: Optional[bool] = field(
+        default=True,
+        metadata={
+            "help": "Whether to use one of the fast tokenizer (backed by the tokenizers library) or not."
+        },
+    )
+    revision: Optional[str] = field(
```
Comment (naren-determined) [inline model_hub/model_hub/huggingface/_arg_parser.py:94]: Ditto revision.

File: model_hub/model_hub/huggingface/_arg_parser.py [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+@dataclass
+class ModelKwargs:
+    pretrained_model_name_or_path: str = field(default=None)
+    cache_dir: Optional[str] = field(
+        default=None,
+        metadata={
+            "help": "Where do you want to store the pretrained models downloaded from huggingface.co"
+        },
+    )
+    revision: Optional[str] = field(
```
Comment (naren-determined) [inline model_hub/model_hub/huggingface/_arg_parser.py:118]: Ditto revision.

File: model_hub/model_hub/huggingface/_trial.py [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+            "weight_decay": weight_decay,
+        },
+        {
+            "params": [p for n, p in model.named_parameters() if any(nd in n for nd in no_decay)],
+            "weight_decay": 0.0,
+        },
+    ]
+
+
+def build_default_optimizer(
```
Comment (naren-determined) [inline model_hub/model_hub/huggingface/_trial.py:108]: Does Huggingface only use Adafactor or AdamW?

File: model_hub/model_hub/huggingface/_trial.py [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+
+        if self.hparams.use_apex_amp:
+            self.model, self.optimizer = self.context.configure_apex_amp(
+                models=self.model,
+                optimizers=self.optimizer,
+            )
+
+        self.lr_scheduler = self.context.wrap_lr_scheduler(
+            build_default_lr_scheduler(self.optimizer, scheduler_args),
+            LRScheduler.StepMode.STEP_EVERY_BATCH,
```
Comment (naren-determined) [inline model_hub/model_hub/huggingface/_trial.py:241]: Do we want to allow users to set this StepMode or does Huggingface not have that abstraction?

**Review by naren-determined — CHANGES_REQUESTED** [review]
Directory wise you have `model_hub/model_hub` -- can we rename the second level down to something else?

File: model_hub/examples/huggingface/token-classification/ner_config.yaml [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+  # Training Args
+  global_batch_size: 16
+  learning_rate: 5e-5
+  adam_epsilon: 1e-8
+  weight_decay: 0
+  lr_scheduler_type: linear
+  num_warmup_steps: 750
+  num_training_steps: 1000
+environment:
+    image: liamdetermined/development:model-hub-0.1.0.dev0-cuda-10.1-pytorch-1.7-tf-1.15-gpu-4a93d25
```
[the remark this answers is not in the document]
Comment (liamcli) [reply model_hub/examples/huggingface/token-classification/ner_config.yaml]: the plan is to always set this to the latest environment image for model-hub and the corresponding versions of transformers and datasets that we rely on.  this way, the user knows which version of model-hub their experiment code was written for.

File: model_hub/model_hub/huggingface/_arg_parser.py (hunk 5 above)
Comment (liamcli) [reply model_hub/model_hub/huggingface/_arg_parser.py:52]: this is the name used by HF so i'm hesitant to change it.

File: model_hub/model_hub/huggingface/_trial.py (hunk 8 above)
Comment (liamcli) [reply model_hub/model_hub/huggingface/_trial.py:108]: yes, this matches the behavior of the HF trainer if no optimizer is provided.
https://github.com/huggingface/transformers/blob/master/src/transformers/trainer.py#L616

File: model_hub/model_hub/huggingface/_trial.py (hunk 9 above)
Comment (liamcli) [reply model_hub/model_hub/huggingface/_trial.py:241]: our BaseTransformerTrial will basically replicate the same training loop as Trainer.  

For HF Trainer, all lr schedulers returned from transformer's get_scheduler method step every batch:
https://github.com/huggingface/transformers/blob/master/src/transformers/trainer.py#L1094

**Comment by liamcli** [comment]
@naren-determined addressed your first round of comments here.  
going to close this PR since have a new PR for a fork on determined that allows us to test more functionality.  
new PR: https://github.com/determined-ai/determined/pull/2068

**Pull request closed without merging**